qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,217 @@
+    param, param_nr = get_next_arguments(action, 1)
+    output, _ = get_next_arguments(action, 2)
+
+    non_req = []
+
+    # unravel potential unions
+    input = unravel_union(input)
+    param = unravel_union(param)
+
+    for x in input_nr + param_nr:
+        non_req += unravel_union(x)
+
+    for i in input:
+        for p in param:
+            if non_req:
+                non_req = get_combinations(non_req)
+                for nr in non_req:
+                    d = {}
+                    d['inputs'] = i
+                    d['params'] = p
+                    d['outputs'] = output
+                    d['non_req'] = nr
+                    results.append(d)
+            else:
+                d = {}
+                d['inputs'] = i
+                d['params'] = p
+                d['outputs'] = output
+                d['non_req'] = []
+                results.append(d)
+
+    return results
+
+
+def unravel_union(l):
+    """
+    Unravel Union node to get all permutations of types for each action
+
+    Parameters
+    ----------
+    list : list of Qiime2.types
+
+    Returns
+    -------
+    list of lists - list of permuations of types for each action
+
+    """
+
+    result = [l]
+    for i, x in enumerate(l):
+        if 'members' in x.__dict__ and len(x.__dict__['members']) > 0:
+            members = list(x.__dict__['members'])
+            temp = copy.deepcopy(result)
+
+            # update result with first element of types in member
+            for each_list in result:
+                each_list[i] = members[0]
+
+            # add in other permutations of types in member
+            for n in range(1, len(members)):
+                copy_result = copy.deepcopy(temp)
+                for each_list in copy_result:
+                    each_list[i] = members[n]
+                result += copy_result
+    return result
+
+
+def build_graph(sigs=[]):
+    """
+    Constructs a networkx graph with different semantic types
+    and actions as nodes
+
+    Parameters
+    ----------
+    sigs : list of strings
+
+    Returns
+    -------
+    nx.DiGraph - networkx graph connected based on all or specified methods
+
+    """
+
+    G = nx.DiGraph()
+    action_list = []
+
+    # get all actions or specifc actions if specified in sigs
+    pm = qiime2.sdk.PluginManager()
+    if not sigs:
+        for pgn, pg in pm.plugins.items():
+            action_list += list(pg.actions.values())
+    else:
+        for pgn, pg in pm.plugins.items():
+            for action in pg.actions.keys():
+                if str(action) in sigs:
+                    action_list.append(pg.actions[str(action)])
+
+    for action in action_list:
+        node_combs = generate_nodes_by_action(action)
+
+        # each action node is a str(dict)
+        # each Qiime2.type node is the type itself
+        for dict in node_combs:
+            if not G.has_node(str(dict)):
+                G.add_node(str(dict), value=action)
```

Comment:
Ok, after looking at the networkx docs, I understand why you are using this as the key.

It is a reasonable workaround, however, another approach would be to use `id(dict)` which is the object identity (in CPython the memory address). I think that's a reasonable thing to do as looking up specific nodes isn't as useful as looking up by `value`.

Reply:
Not sure why id() isn't returning the same thing when the dictionaries created are exactly the same (could be because id == memory location of that object?) 

Would you still like to use ids but store the dictionaries as a value and somehow check if that node exists if the dictionaries match?